src/algebra/module/injective.lean

```diff
@@ -0,0 +1,591 @@
+  is an `R`-linear map `h : Y ⟶ Q` such that `g = h ∘ f`
+  ```
+  X --- f ---> Y
+  |
+  | g
+  v
+  Q
+  ```
+* `module.Baer`: an `R`-module `Q` satisfies Baer's criterion if any `R`-linear map from an
+  `ideal R` extends to an `R`-linear map `R ⟶ Q`
+
+## Main statements
+
+* `module.Baer.criterion`: an `R`-module is injective if it is Baer.
+
+-/
+
+
+noncomputable theory
+
+universes u v
+
+variables (R : Type u) [ring R] (Q : Type (max u v)) [add_comm_group Q] [module R Q]
+
+/--An `R`-module `Q` is injective if and only if every injective `R`-linear map descends to a linear
+map to `Q`, i.e. in the following diagram, if `f` is injective then there is an `R`-linear map
+`h : Y ⟶ Q` such that `g = h ∘ f`
+  ```
+  X --- f ---> Y
+  |
+  | g
+  v
+  Q
+  ```
+-/
+class module.injective : Prop :=
+(out : ∀ (X Y : Type (max u v)) [add_comm_group X] [add_comm_group Y] [module R X] [module R Y]
+  (f : X →ₗ[R] Y) (hf : function.injective f) (g : X →ₗ[R] Q),
+  ∃ (h : Y →ₗ[R] Q), ∀ x, h (f x) = g x)
+
+lemma module.injective_object_of_injective_module [module.injective.{u v} R Q] :
+  category_theory.injective.{max u v} (⟨Q⟩ : Module.{max u v} R) :=
+{ factors := λ X Y g f mn, begin
+    rcases module.injective.out X Y f ((Module.mono_iff_injective f).mp mn) g with ⟨h, eq1⟩,
+    exact ⟨h, linear_map.ext eq1⟩,
+  end }
+
+lemma module.injective_module_of_injective_object
+  [category_theory.injective.{max u v} (⟨Q⟩ : Module.{max u v} R)] :
+  module.injective.{u v} R Q :=
+{ out := λ X Y ins1 ins2 ins3 ins4 f hf g, begin
+    resetI,
+    rcases @category_theory.injective.factors (Module R) _ ⟨Q⟩ _ ⟨X⟩ ⟨Y⟩ g f
+      ((Module.mono_iff_injective _).mpr hf) with ⟨h, eq1⟩,
+    refine ⟨h, λ x, by rw ← eq1; refl⟩,
+  end }
+
+lemma module.injective_iff_injective_object :
+  module.injective.{u v} R Q ↔ category_theory.injective.{max u v} (⟨Q⟩ : Module.{max u v} R) :=
+⟨λ h, @@module.injective_object_of_injective_module R _ Q _ _ h,
+  λ h, @@module.injective_module_of_injective_object R _ Q _ _ h⟩
+
+/--An `R`-module `Q` satisfies Baer's criterion if any `R`-linear map from an `ideal R` extends to
+an `R`-linear map `R ⟶ Q`-/
+def module.Baer : Prop := ∀ (I : ideal R) (g : I →ₗ[R] Q), ∃ (g' : R →ₗ[R] Q),
+  ∀ (x : R) (mem : x ∈ I), g' x = g ⟨x, mem⟩
+
+namespace Baer
+
+variables {R} {M N : Type (max u v)} [add_comm_group M] [add_comm_group N]
+variables [module R M] [module R N] {i : M →ₗ[R] N} (hi : function.injective i) (f : M →ₗ[R] Q)
+variable {Q}
+include hi
+
+/-- If we view `M` as a submodule of `N` via the injective linear map `i : M ↪ N`, then a submodule
+between `M` and `N` is a submodule `N'` of `N`. To prove Baer's criterion, we need to consider
+pairs of `(N', f')` such that `M ≤ N' ≤ N` and `f'` extends `f`. -/
+structure extension_of extends submodule R N :=
+(le : i.range ≤ to_submodule)
+(extension : to_submodule →ₗ[R] Q)
+(is_extension: ∀ (m : M), f m = extension ⟨i m, le ⟨m, rfl⟩⟩)
```

Comment:
I'm pretty sure the Zorn's lemma part can be simplified significantly by using `linear_pmap`, in particular the already defined lattice structure:
```suggestion
structure extension_of extends linear_pmap R N Q :=
(le : i.range ≤ domain)
(is_extension: ∀ (m : M), f m = to_fun ⟨i m, le ⟨m, rfl⟩⟩)
```

Reply:
thanks for the review, I am in the process of rewriting this pr using this suggestion

Reply:
As you have predicted, the Zorn's lemma part has now been significantly simplified. @Vierkantor